tensorboard/components/tensor_widget/selection.ts

```diff
@@ -0,0 +1,340 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {size} from './shape-utils';
+import {Shape, TensorViewSlicingSpec} from './types';
+
+/**
+ * The possible status of a selected cell.
+ */
+export enum CellSelectionStatus {
+  SELECTED = 1,
+  LEFT_EDGE,
+  RIGHT_EDGE,
+  TOP_EDGE,
+  BOTTOM_EDGE,
+}
+
+/**
+ * Possible directions of selection movement.
+ */
+export enum SelectionMoveDirection {
+  UP = 1,
+  DOWN,
+  LEFT,
+  RIGHT,
+}
+
+/**
+ * The selection state within a n-dimensional tensor.
+ *
+ * This class keeps track of what element(s) are selected in the
+ * current viewing dimensions of the tensor. It provides capabiliities to:
+ * - Query whether a given set of indices falls into the selection, and
+ * - if so, whether it belongs to any of the four edges of selection.
+ * - When a selection is moved, what the new selection is and
+ * - if the new selection falls out of the current slicing spec, how the
+ *   slicing spec ought to be updated to accommodate it.
+ */
+export class TensorElementSelection {
+  private sliceDims: number[] = [];
+  private sliceIndices: number[] = [];
+  private viewDims: number[];
+  private rowStart: number;
+  private colStart: number;
+  private rowCount: number;
+  private colCount: number;
+
+  /**
+   * Constructor of TensorElementSelection
+   * @param shape Shape of the tensor in which the viewing and selection is
+   *   taking place.
+   * @param slicingSpec The current slicing spec for the tensor.
+   * @param rowStart The starting row of selection, within the indices
+   *   framework of the original tensor (i.e., *not* with respect to the
+   *   slicing spec.)
+   * @param colStart The starting column of selection, within the indices
+   *   framework of the original tensor (i.e., *not* with respect to the
+   *   slicing spec.)
+   * @param rowCount How many rows are selected.
+   * @param colCount How many columns are selected.
+   */
+  constructor(
+    private readonly shape: Shape,
+    readonly slicingSpec: TensorViewSlicingSpec,
+    rowStart?: number,
+    colStart?: number,
+    rowCount?: number,
+    colCount?: number
+  ) {
+    if (size(this.shape) === 0) {
+      throw new Error(
+        `TensorElementSelection doesn't support tensor with zero elements.`
```

Comment:
Doesn't `(size(this.shape)) === 0)` imply that the tensor is a scalar, not that the tensor has zero elements?

Reply:
`size(this.shape) === 0` doesn't mean this is a scalar. It means you have a shape such as [0, 10], or [3, 0, 5]. A scalar has a `size()` of 1.